services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java

```diff
@@ -31,12 +48,105 @@
+            // we need to clone the key first, to not change the kid of the original key so that the next request still can find it.
+            signingKey = signingKey.cloneKey();
+            signingKey.setKid(keyId);
+        }
+        kid.ifPresent(signingKey::setKid);
+        SignatureProvider signatureProvider = keycloakSession.getProvider(SignatureProvider.class, algorithmType);
+        signatureSignerContext = signatureProvider.signer(signingKey);
+
+        LOGGER.debugf("Successfully initiated the SD-JWT Signing Service with algorithm %s.", algorithmType);
     }
 
     @Override
     public String signCredential(VerifiableCredential verifiableCredential) {
-        throw new UnsupportedOperationException("SD-JWT Signing is not yet supported.");
+
+        DisclosureSpec.Builder disclosureSpecBuilder = DisclosureSpec.builder();
+        CredentialSubject credentialSubject = verifiableCredential.getCredentialSubject();
+        JsonNode claimSet = objectMapper.valueToTree(credentialSubject);
+        // put all claims into the disclosure spec, except the one to be kept visible
+        credentialSubject.getClaims()
+                .entrySet()
+                .stream()
+                .filter(entry -> !visibleClaims.contains(entry.getKey()))
+                .forEach(entry -> {
+                    if (entry instanceof List<?> listValue) {
+                        IntStream.range(0, listValue.size())
+                                .forEach(i -> disclosureSpecBuilder.withUndisclosedArrayElt(entry.getKey(), i, SdJwtUtils.randomSalt()));
+                    } else {
+                        disclosureSpecBuilder.withUndisclosedClaim(entry.getKey(), SdJwtUtils.randomSalt());
+                    }
+                });
+
+        // add the configured number of decoys
+        if (decoys != 0) {
+            IntStream.range(0, decoys)
+                    .forEach(i -> disclosureSpecBuilder.withDecoyClaim(SdJwtUtils.randomSalt()));
+        }
+
+        ObjectNode rootNode = claimSet.withObject("");
+        rootNode.put(ISSUER_CLAIM, issuerDid);
+
+        // Get the issuance date from the credential. Since nbf is mandatory, we set it to the current time if not
+        // provided
+        long iat = Optional.ofNullable(verifiableCredential.getIssuanceDate())
+                .map(issuanceDate -> issuanceDate.toInstant().getEpochSecond())
+                .orElse((long) timeProvider.currentTimeSeconds());
+        rootNode.put(NOT_BEFORE_CLAIM, iat);
+        if (verifiableCredential.getType() == null || verifiableCredential.getType().size() != 1) {
+            throw new SigningServiceException("SD-JWT only supports single type credentials.");
+        }
+        rootNode.put(VERIFIABLE_CREDENTIAL_TYPE_CLAIM, verifiableCredential.getType().get(0));
+        rootNode.put(SELECTIVE_DISCLOSURE_ALGORITHM_CLAIM, hashAlgorithm);
+        rootNode.put(CREDENTIAL_ID_CLAIM, JwtSigningService.createCredentialId(verifiableCredential));
```

Comment:
The algorithm claim is added by the SdJwt if necessary. line 
```
        rootNode.put(SELECTIVE_DISCLOSURE_ALGORITHM_CLAIM, hashAlgorithm); 
```
can be left removed.

```suggestion
        rootNode.put(CREDENTIAL_ID_CLAIM, JwtSigningService.createCredentialId(verifiableCredential));
```